### PR TITLE
importccl: use correct targeted column ordering

### LINKED ISSUE #45422: IMPORT INTO does not seem to respect named column order
**Describe the problem**

```sql
CREATE TABLE Foo (
    id STRING NOT NULL,
    num INT NOT NULL
);
```

and with a csv file `reversed_columns.csv` that contains the following:

```csv
17,hello
18,goodbye
```

Start a web server in the same directory as the csv file, then try to import the data but with the order of the columns reversed as compared to the DDL:

```sql
IMPORT INTO Foo (num, id) CSV DATA ('http://localhost:9001/reversed_columns.csv');
```

This results in an error that makes it clear it's trying to import the data into columns in the same order as they exist in the database rather than as named in the `IMPORT` command:

```
pq: "http://localhost:9001/test.csv": row 1: parse "num" as INT8: could not parse "hello" as type int: strconv.ParseInt: parsing "hello": invalid syntax
```

**Environment:**
 - CockroachDB version: 19.2.4
 - Server OS: Windows 10 1909
 - Client app: `cockroach sql`

### PR DESCRIPTION
**Pull request by C0rWin**
Fixes #45422.

Currently during the import the order of the columns during data
transformation based on the table description rather than order of the
columns as appeared in the import expression.

This commit takes care that during the import from the external data
sourced it will also consider the order of the mentioned columns such
that it will properly apply conversion for accurate data type.

Signed-off-by: Artem Barger <bartem@il.ibm.com>

Release note: none

### CODE REVIEW

**Comment by C0rWin** [comment]
will take care to add UT as well to cover the fix.

**Comment by miretskiy** [comment]
Thanks for your contributions.  Do you mind adding a test case to e.g. import_stmt_test.go?
TestImportIntoCSV is a good candidate.

**Comment by C0rWin** [comment]
> Thanks for your contributions. Do you mind adding a test case to e.g. import_stmt_test.go?
> TestImportIntoCSV is a good candidate.

definitely will do :)

**Comment by miretskiy** [comment]
GitHub CI have been flaky recently.  You might have to rerun those snapshot tests few times.
Again, thanks for doing this

**Review by miretskiy — APPROVED** [review]

File: pkg/ccl/importccl/import_stmt_test.go [hunk 1]
```diff
@@ -2127,6 +2127,17 @@ func TestImportIntoCSV(t *testing.T) {
 		)
 	})
 
+	// Tests the case where we create table columns in specific order while trying
+	// to import data from csv where columns order is different and import expression
+	// defines in what order columns should be imported to align with table definition
+	t.Run("target-cols-reordered", func(t *testing.T) {
+		sqlDB.Exec(t, "CREATE TABLE t (a STRING NOT NULL, b INT NOT NULL)")
+		defer sqlDB.Exec(t, `DROP TABLE t`)
+
```
Comment (dt) [inline pkg/ccl/importccl/import_stmt_test.go]: i'm inclined to also include specifying a strict subset of the columns to ensure not only that we're ordering the set correctly but also using the right set. maybe something like 
```
 		sqlDB.Exec(t, "CREATE TABLE t (a INT PRIMARY KEY, b INT, c STRING)")
 		defer sqlDB.Exec(t, `DROP TABLE t`)
    data := `"c is a string", 1`

		srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if r.Method == "GET" {
				_, _ = w.Write([]byte(data))
			}
		}))
		defer srv.Close()
```
and then swap `(b, a)` for `(c, a)` and `testFiles.files[0]` for `srv.URL` below?

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45713)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @C0rWin)

___
*[pkg/ccl/importccl/import_stmt_test.go, line 2042 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45713#-M1bh1nA7TEjYPXQx1P1:-M1bh1nBCj02DXR2EzPF:bc1abe1) ([raw file](https://github.com/cockroachdb/cockroach/blob/42deeb2995b1d56d540a53c5b37853a1201e6811/pkg/ccl/importccl/import_stmt_test.go#L2042)):*
> ```Go
> 			defer sqlDB.Exec(t, `DROP TABLE t`)
> 
> 			sqlDB.Exec(t, `IMPORT INTO t (a) CSV DATA ($1)`, srv.URL)
> ```

@miretskiy huh, is this unexpected? `data` is a CSV with 6 fields and we're importing into 1 column -- don't we expect a "got 6 fields, expected 1" error?

**Review by C0rWin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45713)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @C0rWin and @dt)

___
*[pkg/ccl/importccl/import_stmt_test.go, line 2136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45713#-M1bMTmD6Z1lCXr8xTui-r2-2136:-M1biKAG5MI3fX7PRXm0:b9szk38) ([raw file](https://github.com/cockroachdb/cockroach/blob/42deeb2995b1d56d540a53c5b37853a1201e6811/pkg/ccl/importccl/import_stmt_test.go#L2136)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

i'm inclined to also include specifying a strict subset of the columns to ensure not only that we're ordering the set correctly but also using the right set, ideally including an unspecified column right in the middle of some specified ones. Maybe something like this:
```
 		sqlDB.Exec(t, "CREATE TABLE t (a INT PRIMARY KEY, b INT, c STRING)")
 		defer sqlDB.Exec(t, `DROP TABLE t`)

    const data = `"c is a string",1\n"another string",2`

		srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if r.Method == "GET" {
				_, _ = w.Write([]byte(data))
			}
		}))
		defer srv.Close()
```
and then change `(b, a)` to `(c, a)` and `testFiles.files[0]` to `srv.URL` below?
</blockquote></details>

certainly, thanks for suggestion, will do it.

File: pkg/ccl/importccl/import_stmt_test.go (hunk 1 above)
Comment (C0rWin) [reply pkg/ccl/importccl/import_stmt_test.go]: done

**Comment by miretskiy** [comment]
didn't we specify column a as the only column we're importing?

**Comment by C0rWin** [comment]
Submitted another one instead, PR #45747

**Pull request closed without merging**